From 57bd79561caa26497cfa74038797372f2819b4f2 Mon Sep 17 00:00:00 2001
From: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
Date: Mon, 7 Oct 2013 17:47:40 +0300
Subject: [PORT FROM MAIN] Telephony: Call fetchSimRecords only when imsi read is done

BZ: 143292

With LTE enabled device, device takes more than 30seconds
to register to network.

With LTE enabled device, default APN needs to be configured
before the network selection request is sent to modem. Default
APN information is retrieved based on the IMSI. Since IMSI read
is delayed due to quite a lot of SIM file read requests issued on
sim ready, network selection request to modem is also delayed.

Upon SIM Ready, read the EFad and EFimsi files first. Once
the EFad and EFimsi files are read, other SIM files will be
read.

Change-Id: Icfeace741eada6181072f585cd88259a04b664b2
Orig-Change-Id: I1ffb8c17fb7ec0cc088af9b2f1c127b1dceadd29
Category: aosp improvement
Domain: Telephony-Telephony-FWK
Origin: internal
Upstream-Candidate: yes
Signed-off-by: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
---
 .../internal/telephony/uicc/SIMRecords.java        |   24 ++++++++++++++-----
 1 files changed, 17 insertions(+), 7 deletions(-)

diff --git a/src/java/com/android/internal/telephony/uicc/SIMRecords.java b/src/java/com/android/internal/telephony/uicc/SIMRecords.java
index f532531..9d23356 100644
--- a/src/java/com/android/internal/telephony/uicc/SIMRecords.java
+++ b/src/java/com/android/internal/telephony/uicc/SIMRecords.java
@@ -538,8 +538,8 @@ public class SIMRecords extends IccRecords {
         if (fileChanged) {
             // A future optimization would be to inspect fileList and
             // only reload those files that we care about.  For now,
-            // just re-fetch all SIM records that we cache.
-            fetchSimRecords();
+            // just re-fetch all SIM records that we cache starting with EFad.
+            getEFad();
         }
     }
 
@@ -587,6 +587,9 @@ public class SIMRecords extends IccRecords {
             case EVENT_GET_IMSI_DONE:
                 isRecordLoadResponse = true;
 
+                // Now fetch all the SIM files.
+                fetchSimRecords();
+
                 ar = (AsyncResult)msg.obj;
 
                 if (ar.exception != null) {
@@ -1205,7 +1208,7 @@ public class SIMRecords extends IccRecords {
                 // voicemail number.
                 // TODO: Handle other cases, instead of fetching all.
                 mAdnCache.reset();
-                fetchSimRecords();
+                getEFad();
                 break;
         }
     }
@@ -1379,7 +1382,17 @@ public class SIMRecords extends IccRecords {
 
     @Override
     public void onReady() {
-        fetchSimRecords();
+        getEFad();
+    }
+
+    protected void getEFad() {
+        /*
+         * Upon sim refresh or sim ready, EFad and EFimsi files will be read first.
+         * Once the EFad and EFimsi files are read, other SIM files will be read.
+         */
+        mRecordsRequested = true;
+        mFh.loadEFTransparent(EF_AD, obtainMessage(EVENT_GET_AD_DONE));
+        mRecordsToLoad++;
     }
 
     protected void fetchSimRecords() {
@@ -1387,9 +1400,6 @@ public class SIMRecords extends IccRecords {
 
         if (DBG) log("fetchSimRecords " + mRecordsToLoad);
 
-        mFh.loadEFTransparent(EF_AD, obtainMessage(EVENT_GET_AD_DONE));
-        mRecordsToLoad++;
-
         mFh.loadEFTransparent(EF_ICCID, obtainMessage(EVENT_GET_ICCID_DONE));
         mRecordsToLoad++;
 
-- 
1.7.4.1

